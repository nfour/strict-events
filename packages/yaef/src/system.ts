import { map, reduce } from 'bluebird';

import { ErrorFromCallPoint } from './lib';
import { Omit, UnionToIntersection } from './types';

export function Component<
  E extends IComponentSignature
> (
  input: E,
  callback: (mediator: SimpleMediator<EventTuplesToUnion<E>>) => void,
): IComponent<E> {
  const component = <IComponent<E>> ((mediator) => callback(mediator));

  type ComponentPropConstraints = { [K in keyof IComponent<any>]: any };

  Object.defineProperties(component, <ComponentPropConstraints> {
    name: { value: input.name, writable: false },
    observations: { value: input.observations, writable: false },
    publications: { value: input.publications, writable: false },
    disconnect: { value: () => { /**/ } }, // TODO: make this useful?
  });

  return component;
}

/** This is gay */
const primitiveNames = ['Object', 'Array', 'String', 'Number'];

const CheckMediatorEventName = (from: string) => (name?: string) => {
  if (!name || primitiveNames.includes(name)) {
    throw ErrorFromCallPoint({ fromStackPosition: 3 })(
      `[${from}] Cannot observe event, missing unique 'name' property. Got value: ${name}`,
    );
  }
};

const getNameFromEvent = (event: IEventShapes) => event.name || event.constructor.name;

export type IMediatorEventCallback = (event: any) => any;

export class SimpleMediator<Events extends IEventSignatures> implements IMediator<Events> {
  /** For storing type information */
  Events!: Events;

  observers: Map<any, Array<{ event: any, callback: IMediatorEventCallback }>> = new Map();

  /** Add many observers, say, from another mediator */
  addObservers (observers: SimpleMediator<any>['observers']) {
    this.observers = new Map([
      ...this.observers.entries(),
      ...observers.entries(),
    ]);
  }

  removeObserver (inputCallback: IMediatorEventCallback) {
    for (const observers of this.observers.values()) {
      const indexMatch = observers.findIndex(({ callback }) => callback === inputCallback);

      if (indexMatch >= 0) {
        // Remove it.
        return observers.splice(indexMatch, 1);
      }
    }
  }

  observe <Es extends this['Events']['observations']> (event: Es, callback: (event: Es) => any) {
    const name = event.name;

    CheckMediatorEventName(this.constructor.name)(name);

    /** Set default if not defined */
    if (!this.observers.has(name)) { this.observers.set(name, []); }

    const observers = this.observers.get(name)!;

    // TODO: need to be able to add an observe to a specific placement
    this.observers.set(name, [{ event, callback }, ...observers]);
  }

  publish<Es extends this['Events']['publications']> (
    event: Es,
    /** This name thing exists because we are being cheeky with props */
    payload?: Omit<{ [K in keyof typeof event]: typeof event[K] }, 'name'>,
  ): Es | undefined | any {
    const name = getNameFromEvent(event as IEventShapes);

    CheckMediatorEventName(this.constructor.name)(name);

    if (!this.observers.has(name)) { return; }

    const observers = this.observers.get(name)!;

    if (!observers.length) { return; }

    const firstEvent = observers[0].callback(event);

    if (observers.length <= 1) { return firstEvent; }

    return observers.slice(1).reduce((prevEvent, { callback }) => callback(prevEvent), firstEvent);
  }
}

export class PromisingMediator<Events extends IEventSignatures> extends SimpleMediator<Events> {
  async publish<Es extends this['Events']['publications']> (
    event: Es,
    /** This name thing exists because we are being cheeky with props */
    payload?: Omit<{ [K in keyof typeof event]: typeof event[K] }, 'name'>,
  ): Promise<Es | undefined> {
    const name = getNameFromEvent(event as IEventShapes);

    CheckMediatorEventName(this.constructor.name)(name);

    if (!this.observers.has(name)) { return; }

    const observers = this.observers.get(name)!;

    if (!observers.length) { return; }

    const firstEvent = await observers[0].callback(event);

    if (observers.length <= 1) { return firstEvent; }

    return reduce(observers.slice(1), async (prevEvent, { callback }) => callback(prevEvent), firstEvent);
  }
}

/**
 * TODO: Support `Mediator` property, so one can construct from arbitrary IMediator conformant interfaces
 */
export function ComponentMediator<C extends IComponent<any>> ({ components }: { components: C[] }) {
  /** This should be part of the input as well */
  const mediator = new SimpleMediator<MergeComponentEvents<C>>();

  return {
    mediator,
    async connect () {
      // TODO: if anything tries to .publish before this is called, they should get rekt
      // Or it should be buffered, like in `reaco`

      await map(components, (component) => component(mediator));

      return mediator;
    },
    async disconnect () {
      await map(components, (component) => component.disconnect());
    },
  };
}

export function ComponentSignature<
  In extends Omit<IComponentSignature<any>, 'name'>,
  N extends string = string
> (name: N, input?: In) {
  return { name, ...input || {} } as { name: N } & In;
}

export function EventSignature<In extends Omit<IEventSignature, 'name'>, N extends string = string> (name: N, input?: In) {
  return { name, ...input || {} } as Omit<In, 'name'> & { name: N };
}

export interface IMediator<Events extends IEventSignatures> {
  Events: Events;

  observe <Es extends this['Events']['observations']> (event: Es, ...args: any[]): any;
  publish <Es extends this['Events']['publications']> (event: Es, payload?: unknown): any;
}

export interface MergeComponentEvents<C extends IComponent<any>> {
  observations: C['observations'][number] | C['publications'][number];
  publications: C['observations'][number] | C['publications'][number];
}

export interface EventTuplesToUnion<E extends IComponentEvents> {
  observations: E['observations'][number];
  publications: E['publications'][number];
}

// tslint:disable-next-line: interface-over-type-literal
export type IEventSignature = { readonly name?: string, [k: string]: unknown };

export type IEventShapes = UnionToIntersection<IEventSignature>;

export interface IEventSignatures {
  observations: IEventSignature;
  publications: IEventSignature;
}

export interface IComponentSignature<E extends IEventSignature = IEventSignature> {
  name: string;
  observations: E[];
  publications: E[];
}

export interface IComponentEvents<Event extends IEventSignature = IEventSignature> {
  observations: Event[];
  publications: Event[];
}

export type AnonComponent<Events extends IEventSignatures> = (mediator: SimpleMediator<Events>) => void;

/**
 * TODO: Need `In['name']` to actually be a literal instead of `string`
 */
export interface IComponent<
  In extends IComponentSignature = IComponentSignature,
  M = SimpleMediator<EventTuplesToUnion<In>>, // TODO: support arbitrary mediator signatures, dont couple to Simple
> {
  (mediator: M): void | Promise<void>;

  name: In['name'];

  observations: In['observations'];
  publications: In['publications'];

  disconnect (): void | Promise<void>;
}
