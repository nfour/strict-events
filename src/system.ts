
export type IEvent = unknown;

export interface IComponentEvents {
  observations: IEvent;
  publications: IEvent;
}

export type RefineEventInput<E extends IComponentEvents> = (
  {
    observations: E['observations'][number];
    publications: E['publications'][number];
  }
);

export type AnonComponent<Events extends IComponentEvents> = (mediator: IMediator<RefineEventInput<Events>>) => void;

export interface IComponent<Events extends IComponentEvents> {
  readonly name?: string;
  // configuration: Configuration; ???

  (mediator: IMediator<RefineEventInput<Events>>): void;

  observations: Events['observations'];
  publications: Events['publications'];
}

export function Component<E extends IComponentEvents> (
  eventInput: E,
  cb: (mediator: IMediator<RefineEventInput<E>>) => void): IComponent<E> {
  // ...

  return {} as any;
}

/** TODO: */
export type IEventReturn<E extends IEvent> = E; // TODO: stream

export interface IMediatedEvents {
  observations: IEvent;
  publications: IEvent;
}

export interface IMediator<Events extends IMediatedEvents> {
  _events: Events;

  observe: <Es extends this['_events']['observations']> (event: Es, cb?: (event: Es) => any) => IEventReturn<Es>;
  publish: <Es extends this['_events']['publications']> (event: Es) => IEventReturn<Es>;
}
