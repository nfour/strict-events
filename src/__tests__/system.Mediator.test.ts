import { delay } from 'bluebird';

import { Component, ComponentMediator } from '../system';

test('Can observe published events', async () => {
  const { mediator } = ComponentMediator<any>({ components: [] });

  expect(() => mediator.publish({}, {})).toThrowErrorMatchingSnapshot();
  expect(() => mediator.observe({}, () => { /** */})).toThrowErrorMatchingSnapshot();
  expect(() => mediator.publish({ name: 'foo' })).not.toThrowError();
  expect(() => mediator.publish(class Foo {})).not.toThrowError();

  const X = { name: 'x' };

  const xEventReceived = jest.fn();

  mediator.observe({ name: 'x' }, () => xEventReceived());
  mediator.publish({ name: 'x' });

  await delay(5); // I am too lazy to controlflow async stuff

  expect(xEventReceived).toBeCalledTimes(1);
});

test('Can mediate events within components', async () => {
  const Foo = { name: 'Foo', a: 1, x: 9 as number } as const;
  class Bar { static b: 1; }

  const eventBarReceived = jest.fn();
  const eventFooReceived = jest.fn();

  const component1 = Component({ observations: [Foo], publications: [Bar] }, (m) => {
    m.observe(Foo, (args) => {
      expect(args.a === 1).toBe(true);

      eventFooReceived();

      m.publish(Bar, { b: 3 } as any);
    });
  });

  const component2 = Component({ observations: [Bar], publications: [] }, (m) => {
    m.observe(Bar, (args) => {
      eventBarReceived();
    });
  });

  const componentMediator = ComponentMediator({ components: [component1, component2] });

  const mediator = componentMediator.initialize();

  mediator.publish(Foo, { a: 1, x: 999 });

  await delay(5);

  expect(eventBarReceived).toBeCalledTimes(1);
  expect(eventFooReceived).toBeCalledTimes(1);
});
