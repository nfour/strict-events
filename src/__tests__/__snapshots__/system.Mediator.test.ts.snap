// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can observe published events 1`] = `"[SimpleMediator] Cannot observe event, missing unique 'name' property. Got value: Object"`;

exports[`Can observe published events 2`] = `"[SimpleMediator] Cannot observe event, missing unique 'name' property. Got value: undefined"`;
